Switch to using UnliftIO.Exception bracket

rvl opened the pull request:
### Issue Number

Probably

### Overview

It turns out that the `Control.Exception.bracket` function isn't very safe to use with async code. The reason is the bracket cleanup handler may be interrupted by async exceptions, and if that happens, then cleanup action may not be done.

I am wondering whether this may be the cause of some test flakiness. Also I wonder whether cleanup handlers are the cause of deadlocking tests.

This change replaces `Control.Exception` with `UnliftIO.Exception` - the "async safe" variant.

The other benefit of `MonadUnliftIO` is that it provides "unlift" for functions of the form `withFoo :: (foo -> IO a) -> IO a` -- i.e. it lets you replace `IO` with other kinds of `m`.

Docs here: https://github.com/fpco/unliftio/tree/master/unliftio#readme

So it may fix flaky tests and/or help to avoid future concurrency bugs.

### Comments

- Most changes are replacing `Control.Exception` with `UnliftIO.Exception` and `Control.Concurrent.Async` with `UnliftIO.Async`.
- The `MonadCatch` class is replaced with `MonadUnliftIO`.
- As mentioned in the UnliftIO readme, we lose the `MonadCatch` instance for `Either` - not a bad thing really.
- The use of `MonadFail` was removed, and that was a good thing.

rvl committed:
Switch to using UnliftIO.Exception bracket

It turns out that the Control.Exception.bracket function isn't very
safe to use with async code. The reason is the bracket cleanup handler
may be interrupted by async exceptions, and if that happens, then
cleanup action may not be done.

I am wondering whether this may be the cause of some test
flakiness. Also I wonder whether cleanup handlers are the cause of
deadlocking tests.

This change replaces Control.Exception with UnliftIO.Exception - the
"async safe" variant.

The other benefit of MonadUnliftIO is that it provides "unlift" for
functions of the form withFoo :: (foo -> IO a) -> IO a -- i.e. it lets
you replace IO with other kinds of m.

Docs here: https://github.com/fpco/unliftio/tree/master/unliftio#readme

So it may fix flaky tests and/or help to avoid future concurrency
bugs.

KtorZ commented on lib/core/src/UnliftIO/Compat.hs:
```diff
@@ -39,3 +40,12 @@ mkRetryHandler
+    -> (e -> m a)
+    -> m a
+    -> m a
+handleIf f h = handle (\e -> if f e then h e else throwIO e)
```
:+1:

rvl committed:
Make integration test retrying compatible with unliftio

UnliftIO.Exception functions only catch synchronous exceptions.

rvl committed:
Adjust Test.Hspec.Extra.aroundAll for unliftio

Adds changes to clarify aroundAll and ensure the async thread always
gets cancelled.

rvl committed:
Replace System.Process with UnliftIO.Process

rvl committed:
Let bracketTracer differentiate between sync and async exceptions

rvl committed:
Convert withBackendProcess to MonadUnliftIO

rvl committed:
Work around stake pools metadata crap

Now that bracket cleanup actions are masked, there is a deadlock
somewhere in the mess. That whole module needs to be reworked. For
now, just disable the section which deadlocks.

rvl committed:
Regenerate nix

rvl committed:
Replace Control.Concurrent.MVar with UnliftIO.MVar

rvl committed:
Replace Control.Concurrent with UnliftIO.Concurrent

rvl committed:
Fix stake pools monitor metadata deadlock

rvl committed:
Replace GHC.Conc with UnliftIO.STM

rvl committed:
Add note about masking in sqlite runQuery

KtorZ commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -220,6 +222,10 @@ startSqliteBackend manualMigration autoMigration tr fp = do
+    -- runSqlConn is guarded with a lock because it's not threadsafe in general.
+    -- It is also masked, so that the SqlBackend state is not corrupted if a
+    -- thread gets cancelled while running a query.
+    -- See: https://github.com/yesodweb/persistent/issues/981
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Logging.hs:
```diff
@@ -131,18 +135,26 @@ data BracketLog
-    res <- action `onException` traceWith tr BracketException
-    traceWith tr BracketFinish
-    pure res
+    withException
```
Ok :+1:

KtorZ commented on lib/launcher/src/Cardano/Launcher.hs:
```diff
@@ -148,59 +156,70 @@ instance Exception ProcessHasExited
   where
+    withCreateProcessMod inner =
+        fmap join $ tryJust spawnPredicate $
+        bracket (createProcess process) (liftIO . cleanupProcess) $
```
Why not stick with: https://hackage.haskell.org/package/unliftio-0.2.13/docs/UnliftIO-Process.html#v:withCreateProcess :thinking: ?

rvl replied:
I need to revert this - it's a partial change.

KtorZ commented on lib/test-utils/src/Test/Hspec/Extra.hs:
```diff
@@ -97,27 +101,32 @@ aroundAll acquire =
-                putMVar resource a
-                await release
-            unlock done
+        pid <- async $ finally (unlock done) $ acquire $ \a -> do
```
:+1:

rvl replied:
Actually, there was a mistake here which broke all tests cleanup. It's so embarrassing, I have force pushed over it.

KtorZ commented on lib/test-utils/src/Test/Hspec/Extra.hs:
```diff
@@ -97,27 +101,32 @@ aroundAll acquire =
+it = itWithCustomTimeout (60*minutes)
   where
-    minute = 60
+    minutes = 10
```
I get this change, but this confuses me. Doesn't matter, but it's interesting to see how brains pick those things up.

rvl replied:
I needed to temporarily reduce the timeout to 1 minute - so in this form you can just edit the minutes variable to set the timeout.

KtorZ commented on lib/test-utils/src/Test/Hspec/Extra.hs:
```diff
@@ -126,28 +135,34 @@ itWithCustomTimeout
+       Left (Just firstException) ->
+           throwIO firstException
+       Right () ->
+           pure ()
```
Hmmmm. Okay. We are loosing the "AsyncCancelled" handler and tests will report as failed when we CTRL-C them but It seems to be for the best.

rvl replied:
This is one of the nice things about unliftio - less things to think about when handling exceptions.
`UnliftIO.Exception.catch` will only catch synchronous exceptions.
Any `AsyncCancelled` or `ThreadKilled` exceptions can bubble up.
The disadvantage is that an MVar is needed here since AsyncCancelled (end of race) is never caught.

KtorZ reviewed (approved):
Ok. Nice work @rvl

rvl committed:
Re-export AsyncCancelled from UnliftIO.Compat

rvl committed:
Cardano.Launcher: Ensure subprocesses are cleaned up

With UnliftIO.Process, seems like cardano-node processes from the
cluster aren't killed properly.

This is a temporary fix until we can switch to typed-process.

rvl committed:
Use UnliftIO.Temporary where possible

rvl committed:
withCluster: re-use some env vars code

rvl committed:
Fix some error paths in integration tests

rvl committed:
Fix file descriptor leak in follow function

The integration tests were leaking file descriptors for the
cardano-node unix socket.

Since UnliftIO.Exception.handle does not catch async exceptions, the
cleanup handlers weren't getting run when follow threads were
cancelled.

This changes follow to use bracket so that the node connection is
always closed.
